streams/src/test/java/org/apache/kafka/streams/kstream/internals/SuppressScenarioTest.java

```diff
@@ -493,7 +493,7 @@ public void shouldSupportFinalResultsForSessionWindows() {
                 drainProducerRecords(driver, "output-raw", STRING_DESERIALIZER, LONG_DESERIALIZER),
                 asList(
                     new KeyValueTimestamp<>("[k1@0/0]", 1L, 0L),
-                    new KeyValueTimestamp<>("[k1@0/0]", null, 1L),
+                    new KeyValueTimestamp<>("[k1@0/0]", null, 0L),
```

Comment:
If we merge two sessions, we use the session-end-timestamp on delete for the smaller session now.

Reply:
Maybe -- I was not really sure about this one. We never discussed how deletes should be handled. If we don't use the session-window end-timestamp, it seem we might reintroduce non-determinism -- not sure.

Reply:
This is a very tricky subject, but it _seems_ like the deletes should "happen" at the same time as the update. This would be the window-end time of the final merged window. In that case, we should actually not do any mutation while we're merging, but instead collect all the stuff to delete, and delete it at the end, while we also issue the update. (Since we wouldn't know the timestamp to use until after the merging is done).

Actually, this would also let us fix https://issues.apache.org/jira/browse/KAFKA-8318 , and opens up the possibility of doing a bulk/batch update to the state store.

Reply:
Not sure if I understand the connection to 8318? Why do we need the merged-windowed end-timestamp to fix it? Also not sure how bulk/batch updates related?

Also curious was others think. \cc @guozhangwang @bbejeck @ableegoldman @cadonna @abbccdda

Reply:
Sure, those are just things that become low-hanging fruit as a result of the refactoring we have to do to fix the time semantics here. Since we won't know the timestamp for "this" update until after computing the merged window, we can't actually do any of the state updates/forwards until after merging, which means that we have a batch of updates available. It also means we're in a position to eliminate the unnecessary tombstone.

But those are just extras. The main point is that we're deleting the old window at time 5 here, not time 0. Introducing time-travel to the aggregation logic seems more likely to create problems than not.

Reply:
Bumping this conversation...

Reply:
I inclined to agree that it seems the delete should happen at the same time as the update, meaning that we use the timestamp when the delete action occurs, but I could be wrong.

Reply:
Hmm, I'm going to be a PITA and change my mind on this. Sorry, @bbejeck ...

After some further reflection, here's what I'm thinking:
Session windows have this lifecycle: creation, multiple updates, and deletion. Creation has always taken place effectively at window-end time, since when a window is created, start==end==record.timestamp.

Previously updates to a window were just assigned the timestamp of the event that caused the update. Let's say you have a count aggregation, and you have some window `[0,5]` with a count of 2, and you get some more input events at times `5, 3, 4`, then your sequence of result updates are `3 at time 5, 4 at time 3, and 5 at time 4`. This is semantically problematic, because the timestamps tell you these results are "out of order" and that the "most recent" count is actually 3 :( .

What @mjsax is proposing here is to "pin" all updates to a window to window-end time. Then, the result updates for our example is just `3 at time 5, 4 at time 5, 5 at time 5`. This is totally fine, since, in the case of identical timestamps, offset order is the tie-breaker. Therefore, the "most recent" count is (correctly) 5.

It's worth noting that in general, we get an equally correct sequence of window updates as long as the update times don't go backwards. Just save this thought for a minute.

Now, we come to deletes. Arguably, the delete is just another kind of update. I think this is where Matthias's head was originally at. The same logic applies, if the delete timestamp is equal to the update timestamps and the create timestamp, then offset order is the tie-breaker. Since we delete the window after the creation and updates, the final state of the window is (correctly) "deleted".

Unlike the create/updates, though, the delete is "caused" by an event with a timestamp after the window end time. This is what was tripping me up. It seems fine to report a window update time as the "high watermark" time of all the window updates so far (in the case of disordered events), but it seems weird to report a window update (in this case, the delete) time as "in the past", from the perspective of the event that caused it. That's why I was thinking that we should use the later timestamp, to indicate that the delete was caused at that later time. This is _also_ correct from a time semantics POV, because it preserves the correct order, that the delete comes the create/updates.

So, this is the punchline, that both approaches result in correct time semantics. The only difference is that using the causing-event timestamp for deletes reflects the provenance of the delete more accurately, but I don't think this fact is actually useful for anything. Given that we're already pinning the create and update timestamps to all be the window-end time, I'm thinking we should go ahead and stick with Matthias's original proposal to go ahead and use the same timestamp for the "final" update (aka, the delete).

Reply:
@vvcephei Thanks for the detailed analysis! I think I agree with you and @mjsax on the approach, just to clarify my understanding further:

Today even if we are not merging two session windows, a single session window's update is treated as a delete followed by an update. I think is what https://issues.apache.org/jira/browse/KAFKA-8318 is reporting about..

Now the logic would become that we use the max(ts, window_end_time) for updates, hence:

1) with an update whose ts is smaller than the current end-window AND larger than the current start-window, the window_start/window_end_time would not change in the update record as well as in the changelog. In this case, we can consider optimizing it by not doing the delete followed by an update (i.e. KAFKA-8318). 

** but practically, with rare out-of-ordering data this would probably give very small perf boost, right?

2) with an update whose ts is larger than current window end time, OR smaller than the window start time, we would update it as an delete of the original record (hence the tombstone ts == the old end-time) and then followed by a put of the new record with the new start / end-time == this record's ts.

3) merging two windows is actually equal to updating the smaller window with a larger end time and updating the larger window with a smaller start time (of course, with a single record).

Is my understanding correct?

Reply:
Hey @guozhangwang , just getting back to this thread...

For 1, yes, I think this is the situation, and I agree with your conclusion (under the assumption that out-of-order data is actually rare)

For 2, almost... for "followed by a put of the new record with the new start / end-time == this record's ts", do you mean the new end-time of the window? That's what we will use as the update timestamp. Note, the new end-time of the window might be the same as the old one, which brings us back to KAFKA-8318.

For 3, I'm afraid I don't follow. Looking at the sequence of updates, it's equivalent to deleting both the original windows and then adding a new one that spans both (and is semantically the merge result). Is this what you meant?

Reply:
Sound correct. Only (3 - merging two window) should be two deletes (of the old windows) and one insert for the new merged window with `[first-window-start-ts,second-window-end-ts]`.

> ** but practically, with rare out-of-ordering data this would probably give very small perf boost, right?

It's not about performance, but just annoying to see unnecessary tombstones IMHO.

Reply:
Yeah for 3) I mean the same as you guys, i.e. we are updating two windows by deleting the old records, but with only a new record. So logically it was like "new window replace old window1" and also "new window replace old window2". 

For 2), the end/start-time of the window will only be the same as the old one if the update record's ts < end-time and ts > start-time. I think we are on the same page @vvcephei